bftengine/src/bcstatetransfer/RVBManager.cpp

```diff
@@ -0,0 +1,711 @@
+                "Add digest for block " << current_rvb_id << " "
+                                        << " ,Digest: " << digest.toString());
+      in_mem_rvt_->addNode(current_rvb_id, digest.get(), BLOCK_DIGEST_SIZE);
+    }
+    ++rvb_nodes_added;
+  }
+  if (rvb_nodes_added > 0) {
+    LOG_INFO(logger_,
+             "Updated RVT (add):" << KVLOG(min_block_id, max_block_id, rvb_nodes_added, as_->getLastBlockNum()));
+  }
+}
+
+BlockId RVBManager::nextRvbBlockId(BlockId block_id) const {
+  uint64_t next_rvb_id = config_.fetchRangeSize * (block_id / config_.fetchRangeSize);
+  if (next_rvb_id < block_id) {
+    next_rvb_id += config_.fetchRangeSize;
+  }
+  return next_rvb_id;
+}
+
+void RVBManager::reportLastAgreedPrunableBlockId(uint64_t lastAgreedPrunableBlockId) {
+  LOG_TRACE(logger_, KVLOG(lastAgreedPrunableBlockId));
+  std::lock_guard<std::mutex> guard(pruned_blocks_digests_mutex_);
+  auto initial_size = pruned_blocks_digests_.size();
+  RVBId start_rvb_id = in_mem_rvt_->getMinRvbId();
+
+  if (start_rvb_id == 0) {
+    // In some cases tree is still not built, we still have to keep the pruned digests
+    start_rvb_id = pruned_blocks_digests_.empty() ? as_->getGenesisBlockNum() : pruned_blocks_digests_.back().first;
+  }
+
+  start_rvb_id = nextRvbBlockId(start_rvb_id);
+  if (lastAgreedPrunableBlockId <= start_rvb_id) {
+    LOG_WARN(logger_, "Inconsistent prune report ignored:" << KVLOG(lastAgreedPrunableBlockId, start_rvb_id));
+    return;
+  }
+
+  uint64_t current_rvb_id = start_rvb_id;
+  int32_t num_digests_added{0};
+  while (current_rvb_id <= lastAgreedPrunableBlockId) {
+    STDigest digest;
+    as_->getPrevDigestFromBlock(current_rvb_id + 1, reinterpret_cast<StateTransferDigest*>(digest.getForUpdate()));
+    pruned_blocks_digests_.push_back(std::make_pair(current_rvb_id, std::move(digest)));
+    current_rvb_id += config_.fetchRangeSize;
+    ++num_digests_added;
+  }
+
+  if (initial_size != pruned_blocks_digests_.size()) {
+    ds_->setPrunedBlocksDigests(pruned_blocks_digests_);
+    LOG_INFO(logger_,
+             num_digests_added
+                 << " pruned block digests saved:"
+                 << KVLOG(start_rvb_id, current_rvb_id, lastAgreedPrunableBlockId, pruned_blocks_digests_.size()));
+  }
+}
+
+std::string RVBManager::getStateOfRvbData() const {
+  auto val = in_mem_rvt_->getRootCurrentValueStr();
+  return val.empty() ? "EMPTY" : val;
+}
+
+bool RVBManager::validate() const {
+  // TODO - consider a pedantic mode, in which we also validate level 1 (end of ST) as well by fetching all block
+  // digests
+  return in_mem_rvt_->validate();
+}
+
+void RVBManager::reset(RvbDataInitialSource inital_source) {
+  LOG_TRACE(logger_, "");
+  in_mem_rvt_->clear();
+  stored_rvb_digests_.clear();
+  stored_rvb_digests_group_ids_.clear();
+  last_checkpoint_desc_.makeZero();
+  rvb_data_source_ = inital_source;
+  // we do not clear the pruned digests
+}
+
+std::string RVBManager::getLogLevel() const {
```

Comment:
Should this be util function at some common place?

Reply:
I agree, but I didn't have time to do it
I want to have a new file called BcStateTran_Utils.hpp/cpp
And put there common functions
I've seen more.

For now, I will add a TODO above it.